### DISCUSSION TITLE
Is there a max number of 4 websocket server clients?

### DISCUSSION THREAD

**Discussion by JackV2020** [mathieucarbou/ESPAsyncWebServer#30] [Q&A]
I set up a websocket server on a Wemos D1 mini pro and 1 simple html based client which only connect and after connection only acts with the server when I click on a button. 
I have some logging in my ino to log connects and disconnects. 
When I start browser windows and connect the first 4 I see logging that they connect. 
When I click buttons everything works just fine. 
As soon as I connect with a 5th browser there is a disconnect of one of the other 4. 
Since the html client starts a reconnect after 0.5 seconds I see the same every 0.5 seconds because every time a connection is lost there are 4 remaining and after 0.5 seconds a new 5th comes in.
So my question is if there is a limit of 4 clients that can connect to the websocket server.
I have plenty of dram free when I connect 4 clients : "DFreeHeap":"22824" , "DHeapFragmentation":"3" , "DMaxFreeBlockSize":"22272" 
When I close any 1 of the browser windows the issue stops.

**Comment by mathieucarbou** [comment]
Hi,

If you look at the code, there is indeed a default limit of 4 or 8 depending if running on 8266 or ESP32. 
This limit can be set through macro:

https://github.com/mathieucarbou/ESPAsyncWebServer/blob/05206417de37751a43e65e4b520ec7149c7d4201/src/AsyncWebSocket.h#L51-L55

This feature is there since the original library but was not documented.

**Reply by JackV2020** [reply]
Hi,

I found this in ESPAsyncWebServer/src/AsyncWebSocket.h

#ifdef ESP32
#define DEFAULT_MAX_WS_CLIENTS 8
#else
#define DEFAULT_MAX_WS_CLIENTS 4
#endif

When I change that to :

#ifndef DEFAULT_MAX_WS_CLIENTS
  #ifdef ESP32
    #define DEFAULT_MAX_WS_CLIENTS 8
  #else
    #define DEFAULT_MAX_WS_CLIENTS 4
  #endif
#endif

and put the the next in my ino I can connect 6 clients on my Wemos which uses the ESP8266 :

#define DEFAULT_MAX_WS_CLIENTS 6

#include <ESPAsyncWebServer.h>

Would you be willing to make that change in ESPAsyncWebServer/src/AsyncWebSocket.h so the max clients can be optionally defined by the user of the library ?


Kind regards,

Jack

**Reply by mathieucarbou** [reply]
> Would you be willing to make that change in ESPAsyncWebServer/src/AsyncWebSocket.h so the max clients can be optionally defined by the user of the library ?

Oh yes of course! I do that now - I forgot to make them customizable. 

As a site note, this is not a blocker because this has to be used with `cleanupClients()`, which takes a maxClients parameter.

**Reply by mathieucarbou** [reply]
I've released v2.10.8

**Comment by JackV2020** [comment]
Thank you,
I've installed v2.10.8, compiled and am happy!
To give you an idea about the app I am creating for Wemos...
I want a seperate WiFi for some Wemos devices and want my Raspberry Pi's on my LAN to be able to communicate with them. This will be possible via the web socket server in a 'central' Wemos napt server.

The app creates an access point and has a settings page for hostname and WiFi credentials of the WiFi it needs to connect to.
It has napt in it so on the config page are also settings for napt ssid, password, subnet and dns server.
It also has ntp in it and uses ESPAsyncWebServer as web server and web socket server for 2 websocket applications.
One to store and retrieve data which is buffered on the 2nd heap and one to give access to gpio's and one . 
The datastore is a very simple one. Clients can create/update variables with values and delete variables. When something is created or changed or deleted all interested clients are updated.
With the other one I can control the mode of 6 GPIO's, read and write them digital or with a pwm signal and when the mode of a gpio is input or input_pullup the clients are updated when there is an intrerrupt. This way I can write html files and work with the 6 gpio's. It also supports a ticker which can be switched on and off and change the interval to update clients with the A0 reading.
Clients are only interested in updates of the data of their application so I keep a map between clients and the 2 web socket applications and the update routine only sends updates to the clients which need to be updated. So something between updating a single client and all clients.
And to get some access to LittleFS I created something I call my LittleFSWeb which consists of a bunch of web services and a web interface. I can upload and download files, create (sub)directories and empty files and edit files directly in the web interface.
![image](https://github.com/mathieucarbou/ESPAsyncWebServer/assets/71119057/f033c1d5-215f-4581-a95f-778054b3a2f6)
When you scroll to the bottom you see the totals :
![image](https://github.com/mathieucarbou/ESPAsyncWebServer/assets/71119057/007308a0-1e67-4c89-a7bb-432fd452a163)
The editor supports  .html, .css, .js, .svg, .json and .txt files. An example of the configuration file for ntp:
![image](https://github.com/mathieucarbou/ESPAsyncWebServer/assets/71119057/4f881774-be77-44bd-b694-0457269d76ea)
The webserver supports and brotli (.br) and gzip (.gz) for html, css and js like index.html.br and mycss.css.gz. (.br files take precedence over .gz files and .gz files take precedence over uncompressed files)
The types html, css and js can be extended with the extra extension .proc and these files are processed by a processor routine which now replaces strings surrounded by 2 %'s with values for the readings of the gpio's and some other things.
After startup it checks if some files exist and when they are not there they are created. These are /index.html, /ConfigNAPT/DHCPReservations.txt, /Configntp/ntp.txt /LittleFSWeb/LittleFSWeb.html and 4 examples, 1 for an .html.proc file, 1 for an html web socket data store example and 1 ino file for a wemos data store example and 1 html to control the gpio's.
I may have missed some things here but when things are ready I will put it in my github [JackV2020 ](https://github.com/JackV2020)
Many thanks again for implementing the change in AsyncWebSocket.h
Jack

**Reply by mathieucarbou** [reply]
Nice project! 
I gave up on using Brotli because there is no support of brotli over http on majority of browers.
How do yo use it ?

**Comment by JackV2020** [comment]
I create gz files with 'gzip -k filename' and .br files with 'brotli filename' and put them on LittleFS. (You can brotli and/or gzip any file type) 
(Code snippets are at the end of this post)
When a page is requested from the webserver it first checks if there is a br file and when that exists and the client is not a mobile client it sends the br file. ( mobile clients do not seem to support brotli as far as I can check)
When there is no br file for the requested file the server checks if there is a gz file and when that exists it sends the gz file ( also to mobiles)
When there is also no gz it sends the uncompressed file ( when it exists)

As example I downloaded the web page /LittleFSWeb/LittleFSWeb.html to my Linux laptop and did 'gzip -k LittleFSWeb.html' and 'brotli LittleFSWeb.html' to produce gz and br files and uploaded these to the /LittleFSWeb directory:
![image](https://github.com/mathieucarbou/ESPAsyncWebServer/assets/71119057/3bde53c0-67f8-4503-ab75-3e7e4a20771e)
(favicon.ico is a file which I create on Linux and upload to LittleFS)
/index.html which is created when it does not exist shows this page:
![image](https://github.com/mathieucarbou/ESPAsyncWebServer/assets/71119057/20d1d061-1c8d-4298-9abe-ef2dcc81afe8)
And the LittleFS button starts /LittleFSWeb/LittleFSWeb.html

When I click the button on my laptop I see the next on the console of the Wemos:
Serve Brotli : /LittleFSWeb/LittleFSWeb.html.br
Serve Uncompressed : /favicon.ico

When I click the button on a mobile I see:
Serve Brotli : /LittleFSWeb/LittleFSWeb.html.gz
Serve Uncompressed : /favicon.ico

Below are code snippets showing how I serve the files.

// At the beginning of the app I have 

AsyncWebServer WebServer(80);

// As demo for the processor function (is the last one below) I create some keys

int key1;
int key2;
int key3;
int key4;

// Internal 'proc_processor_example.html.proc' created as /zExamples/proc_processor.html.proc when file is missing.

const char proc_processor_example[] PROGMEM = R"rawliteral(<!DOCTYPE HTML>
<html lang="en">
<head>
</head>
<body>
<center>
The web server has a function named proc_processor which can replace strings in your html, js or css by actual values.<br>
The standard supported strings with their actual values are:<br><br>
LED_BUILTIN: %LED_BUILTIN%<br>
GPIO2: %GPIO2%<br>
GPIO4: %GPIO4%<br>
GPIO5: %GPIO5%<br>
GPIO12: %GPIO12%<br>
GPIO13: %GPIO13%<br>
GPIO14: %GPIO14%<br>
A0: %A0%<br>
KEY1: %KEY1%<br>
KEY2: %KEY2%<br>
KEY3: %KEY3%<br>
KEY4: %KEY4%<br>
<br>
To use that function you add the extension .proc to the name of your file.<br>
To get a string replacement you put a string between 2 %%'s.
<br>
Example : to get the value of LED_BUILTIN you use %%LED_BUILTIN%% in your html.proc, js.proc or css.proc file<br>
<br>
Notes:<br>
- Any other single %% needs to be replaced by 2 %%'s as you can see when you edit this file.<br>
- KEY1..KEY4 are random numbers which are created just after reboot and can be used to encrypt traffic.<br>
- And of cource you can add your own replacements to the proc_processor function.
</center>
</body>
</html>
)rawliteral";

//
// Function to check if the request comes from a mobile device because brotli does not seem to be supported on mobile
//
bool isMobileClient(AsyncWebServerRequest *request) {
  if (request->hasHeader(F("User-Agent"))) {
    String userAgent = request->header(F("User-Agent"));
    userAgent.toLowerCase();
    // Common substrings in mobile user agents
    if (userAgent.indexOf(F("mobile")) != -1 || userAgent.indexOf(F("android")) != -1 || 
        userAgent.indexOf(F("iphone")) != -1 || userAgent.indexOf(F("ipad")) != -1) {
      return true;
    }
  }
  return false;
}

//
// When serving something I tell what the content type is
//
String getContentType(String path) {
  if (path.endsWith(F(".html")) || path.endsWith(F(".html.gz")) || path.endsWith(F(".html.br"))|| path.endsWith(F(".html.proc"))) {
    return F("text/html");
  } else if (path.endsWith(F(".css")) || path.endsWith(F(".css.gz")) || path.endsWith(F(".css.br")) || path.endsWith(F(".css.proc"))) {
    return F("text/css");
  } else if (path.endsWith(F(".js")) || path.endsWith(F(".js.gz")) || path.endsWith(F(".js.br")) || path.endsWith(F(".js.proc"))) {
    return F("application/javascript");
  } else if (path.endsWith(F(".json"))) {
    return F("application/json");
  } else if (path.endsWith(F(".png"))) {
    return F("image/png");
  } else if (path.endsWith(F(".jpg"))) {
    return F("image/jpeg");
  } else if (path.endsWith(F(".ico"))) {
    return F("image/x-icon");
  } else if (path.endsWith(F(".svg"))) {
    return F("image/svg+xml");
  }
  return F("text/plain");
}

// I have more functions which call WebServer.on and are called by setup()
// This one is called as last

void WebServerConfigure() {

  Serial.println(F("Configuring WebServer ..."));

  WebServer.on("/reboot", HTTP_GET, [](AsyncWebServerRequest *request){ ESP.restart(); });

  WebServer.on("/*", HTTP_GET, [](AsyncWebServerRequest *request){
    String path = request->url();

    // Check if the path contains a "." indicating a file
    if (path.indexOf(F(".")) == -1) { path += F("/");}

    // Check if the path ends with a slash, indicating a directory
    if (path.endsWith(F("/"))) { path += F("index.html"); }

// When using compression make sure your browser supports it.
// Firefox : about:config and look for network.http.accept-encoding and change to : gzip, deflate, br, zstd

    // Check for Brotli compressed version first (Compress your file using 'brotli filename.html' to produce 'filename.html.br')
    if ( (! isMobileClient(request) ) && (LittleFS.exists(path + F(".br"))) ) {
      AsyncWebServerResponse *response = request->beginResponse(LittleFS, path + F(".br"), getContentType(path));
      response->addHeader(F("Content-Encoding"), F("br"));
      Serial.println(F("Serve Brotli : ") + path + F(".br"));
      request->send(response);
    } 

    // Check for Gzip compressed version (Compress your file using 'gzip -k filename.html' to produce 'filename.html.gz')
    else if (LittleFS.exists(path + F(".gz"))) {
      AsyncWebServerResponse *response = request->beginResponse(LittleFS, path + F(".gz"), getContentType(path));
      response->addHeader(F("Content-Encoding"), F("gzip"));
      Serial.println(F("Serve Gzip : ") + path + F(".gz"));
      request->send(response);
    } 

    // Check for uncompressed version
    else if (LittleFS.exists(path)) {
      Serial.println(F("Serve Uncompressed : ") + path);
      if (path.endsWith(F(".proc"))) {
        request->send(LittleFS, path, getContentType(path) , false , proc_processor); // false enables processor | true sends file as template.
      } else {
        request->send(LittleFS, path, getContentType(path));
      }
    } 
    // If no file found, return 404
    else {
      Serial.println(F("File Not Found: ")+path);
      request->send(200, "text/html", F("<script>window.setTimeout(function(){window.location.href = '/';}, 3000);</script><body style='background-color: #E6E6FA;'><center><h1>Not Found!! ")+path);
    }
  });

  // Generate 4 random keys with a value between 1..127 which are used in the proc_processor below.
  srand(micros());
  key1 = rand()%(127)+1;
  key2 = rand()%(127)+1;
  key3 = rand()%(127)+1;
  key4 = rand()%(127)+1;

  if (!LittleFS.exists(F("/zExamples/proc_processor.html.proc"))) {
    File file = LittleFS.open(F("/zExamples/proc_processor.html.proc"), "w");
    file.print(FPSTR(proc_processor_example)); 
    file.close();
  }

}

// ======  Processor for the 'WebServer.on("/*"...' above  

String proc_processor(const String& var) {
  if (var == F("LED_BUILTIN")){ return String(digitalRead(LED_BUILTIN));}
  else if (var == F("GPIO2")) { return String(digitalRead(2));}
  else if (var == F("GPIO4")) { return String(digitalRead(4));}
  else if (var == F("GPIO5")) { return String(digitalRead(5));}
  else if (var == F("GPIO12")){ return String(digitalRead(12));}
  else if (var == F("GPIO13")){ return String(digitalRead(13));}
  else if (var == F("GPIO14")){ return String(digitalRead(14));}
  else if (var == F("A0"))    { return String(analogRead(A0));}
  else if (var == F("KEY1"))  { return String(key1);}
  else if (var == F("KEY2"))  { return String(key2);}
  else if (var == F("KEY3"))  { return String(key3);}
  else if (var == F("KEY4"))  { return String(key4);}
  return String();
}

**Reply by proddy** [reply]
I thought brotli only works on http/2 and HTTP SSL which ESPAsyncWebServer doesn't support? I did get it working using https and https://github.com/hoeken/PsychicHTTP a few months ago, but then I quickly switched back to ESPAsyncWS because the heap usage was horrible (was an esp-idf core problem).

**Comment by mathieucarbou** [comment]
psychicHttp is also spawning 8 workers of 4k each... This is a hungry
framework and not yet on part with ESPAsyncWebServer.

Tried brotli too with esp-dash but brotli was only working with Safari.

All chromium based browser and Mozilla are keeping brotli only for https
connections.

Reason is explains in a chromium discussion where they explain that there
are a lot of mim proxy (ISP or else) that are doing content filtering,
caching and optimization, which are serving back the resources after
applying a gzip compression and consequently setting the content encoding
to gzip.

So for brotli content that they do not understand, they would gzip over an
already-commpressed content and would overwrite the content-encodoing type
to set it to gzip when it was brotli. So the client is losing the
information that it was compressed with brotli.

That's why they only allow it through https.

Mathieu

**Comment by JackV2020** [comment]
Thank you for explaining about the optimization in the network.
I had no idea but I understand that on massive amounts of applications passing text data this saves a lot of bandwidth.
I use 2 browsers on my Ubuntu laptop and they both work fine with brotli.
epiphany supported it right away and I had to do something for Mozilla FireFox as I documented in the code above:
// When using compression make sure your browser supports it.
// Firefox : about:config and look for network.http.accept-encoding and change to : gzip, deflate, br, zstd
I am sure gzip was already in there but I am sure that the br was not in there.
I can not get the brotli working in Firefox on a mobile because I have no idea how to get into the right settings and maybe it is not in Firefox on mobile at all.  So I decided to skip brotli for mobiles in the code and serve gzip files which works.
About access over internet, when I go to anything at home I first open one of my vpns which I run on my Raspberry Pi's and the brotli would go trough the vpn but I have not tested that yet. this way I only have access on my own network. 
Also I will not make my Wemos devices accessible from the Internet so ISP's will not be aware of the content and they will be of no concern. 
I did not try on Windows yet but will do that one day. My wife has Windows on her laptop with Firefox and some other browsers. One that comes with AVG and also the Microsoft Edge and maybe Chrome. Interesting to look into that to see if I can get that working with brotli too.

**Reply by JackV2020** [reply]
brotli support.....I give up....I tried on Windows and I can only get FireFox working. All I have to do is enter the same for network.http.accept-encoding as there is already for network.http.accept-encoding.secure, where the secure must be for https i guess, and it works.
![image](https://github.com/mathieucarbou/ESPAsyncWebServer/assets/71119057/a23a9e94-7eea-434b-af48-d0cc0872463e)
 Even ChatGPT does not know what to do to get it working in Edge. It asks me to check and repair my file system on which the brotli file is located. It asks me to check and recheck my server program even though I explain that it works fine for FireFox. It asks to unpack the brotli file to see if the contents are fine. It asks to use the Network Console and that works fine for normal files and gz files but shows absolutely nothing for the br file. I call it an evening now but tomorrow I will comment out everything in my code concerning brotli. Who knows maybe I can activate it one day.....

**Reply by mathieucarbou** [reply]
> Even ChatGPT does not know what to do to get it working in Edge.

Edge is based on Chromium, like Chrome. They do not support Brotli over http sadly.

**Reply by proddy** [reply]
that's correct
